File: agent-blackbox/upmarto-sdk-ts/src/session.ts
import { createHash } from "node:crypto";
import { realpathSync } from "node:fs";
import { resolve } from "node:path";

export function localDateKey(date = new Date()): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

/**
 * Match Rust `Path::canonicalize()` byte-for-byte before slash/case normalization.
 * On Windows, Rust emits extended-length paths (`\\?\C:\...` → `//?/c:/...`).
 */
function canonicalizeWorkspace(workspace: string): string {
  try {
    const absolute = resolve(workspace);
    let canonical = realpathSync.native(absolute);
    if (process.platform === "win32" && !canonical.startsWith("\\\\?\\")) {
      if (/^[A-Za-z]:[\\/]/.test(canonical)) {
        canonical = `\\\\?\\${canonical}`;
      } else if (canonical.startsWith("\\\\")) {
        canonical = `\\\\?\\UNC\\${canonical.slice(2)}`;
      }
    }
    return canonical;
  } catch {
    return workspace;
  }
}

export function deriveSessionId(workspace: string, dateKey = localDateKey()): string {
  const normalized = canonicalizeWorkspace(workspace).replace(/\\/g, "/").toLowerCase();
  const hash = createHash("sha256")
    .update(`${normalized}:${dateKey}`)
    .digest("hex")
    .slice(0, 32);
  return `${hash.slice(0, 8)}-${hash.slice(8, 12)}-${hash.slice(12, 16)}-${hash.slice(16, 20)}-${hash.slice(20, 32)}`;
}

export function resolveSessionId(workspace: string): string {
  return deriveSessionId(workspace, localDateKey());
}

File: agent-blackbox/upmarto-sdk-rust/src/session.rs
use std::path::Path;

use chrono::Datelike;
use sha2::{Digest, Sha256};

/// Deterministic session id: hash(workspace_path + YYYY-MM-DD).
pub fn derive_session_id(workspace: &Path, date_key: &str) -> String {
    let normalized = workspace
        .canonicalize()
        .unwrap_or_else(|_| workspace.to_path_buf())
        .to_string_lossy()
        .replace('\\', "/")
        .to_lowercase();

    let hash = Sha256::digest(format!("{normalized}:{date_key}").as_bytes());
    let hex: String = hash.iter().map(|b| format!("{b:02x}")).collect();
    let h = &hex[..32];
    format!(
        "{}-{}-{}-{}-{}",
        &h[0..8],
        &h[8..12],
        &h[12..16],
        &h[16..20],
        &h[20..32]
    )
}

pub fn local_date_key() -> String {
    let now = chrono::Local::now();
    format!("{:04}-{:02}-{:02}", now.year(), now.month(), now.day())
}

/// Resolve session id for a workspace (deterministic per day).
pub fn resolve_session(workspace: &Path) -> String {
    derive_session_id(workspace, &local_date_key())
}
